### PR TITLE
implement docker system df

### PR DESCRIPTION
**Pull request by ndeloof**
**- What I did**
upstream https://github.com/rumpl/moby/pull/29

### CODE REVIEW

File: daemon/containerd/service.go [hunk 1]
```diff
@@ -99,12 +104,53 @@ func (i *ImageService) ReleaseLayer(rwlayer layer.RWLayer) error {
 // LayerDiskUsage returns the number of bytes used by layer stores
 // called from disk_usage.go
 func (i *ImageService) LayerDiskUsage(ctx context.Context) (int64, error) {
-	panic("not implemented")
+	ch := i.usage.DoChan("LayerDiskUsage", func() (interface{}, error) {
+		var allLayersSize int64
+		snapshotter := i.client.SnapshotService(containerd.DefaultSnapshotter)
```
Comment (thaJeztah) [inline daemon/containerd/service.go:109]: Make sure to check if code didn't change after the original PR was merged in the fork. For example, on this one; https://github.com/rumpl/moby/blame/e4cd5b9a89f311b9032a1aba5e9c8c1d4cda544e/daemon/containerd/service.go#L111

Code was later updated in https://github.com/rumpl/moby/commit/99655cc01035ff599216449d1108430a7edeed16, but parts of that already upstreamed, so this code needs to be updated accordingly.


What we should probably do, is do another rebase of https://github.com/moby/moby/pull/43902, do conflict resolution on that branch, and then take the commits after rebasing to upstream.

**Review by thaJeztah — CHANGES_REQUESTED** [review]
left a comment (yes, it's tricky because the fork has fallen quite a bit behind with upstream)

File: daemon/containerd/service.go (hunk 1 above)
Comment (ndeloof) [reply daemon/containerd/service.go:109]: indeed, thanks

File: daemon/containerd/service.go [hunk 2]
```diff
@@ -99,12 +104,53 @@ func (i *ImageService) ReleaseLayer(rwlayer layer.RWLayer) error {
...
+		// Get all top images with extra attributes
+		images, err := i.Images(ctx, types.ImageListOptions{
+			Filters:        filters.NewArgs(),
+			SharedSize:     true,
+			ContainerCount: true,
+		})
+		if err != nil {
+			return nil, fmt.Errorf("failed to retrieve image list: %v", err)
+		}
+		return images, nil
```
Comment (thaJeztah) [inline daemon/containerd/service.go:143]: It looks like the existing implementation for graph drivers is exactly the same, and also has the same issue, but I noticed two things here;

- can you rename the `images` variable (as it collides with the `images` import)
- the `fmt.Errorf()` is masking the underlying error-type, which would be a `errdefs.InvalidParam`, so we should import `pkg/errors` and use `errors.Wrap()` here.


```suggestion
		imgs, err := i.Images(ctx, types.ImageListOptions{
			Filters:        filters.NewArgs(),
			SharedSize:     true,
			ContainerCount: true,
		})
		if err != nil {
			return nil, errors.Wrap(err, "failed to retrieve image list")
		}
		return imgs, nil
```

Comment (thaJeztah) [reply daemon/containerd/service.go:143]: Looks like the `images` variable colliding is not an _actual_ issue in the graphdriver implementation, but perhaps it's good to make the same changes in that code as well (as a separate commit) to keep the code the same.

**Review by thaJeztah — CHANGES_REQUESTED** [review]
Left a comment, but otherwise looks sane to me.

For the "graphdriver" implementation, could you (as I commented) make that change as a separate commit?

File: daemon/images/service.go [hunk 3]
```diff
@@ -245,15 +244,15 @@ func (i *ImageService) getLayerRefs() map[layer.ChainID]int {
...
 		// Get all top images with extra attributes
-		images, err := i.Images(ctx, types.ImageListOptions{
+		imgs, err := i.Images(ctx, types.ImageListOptions{
 			Filters:        filters.NewArgs(),
 			SharedSize:     true,
 			ContainerCount: true,
 		})
 		if err != nil {
-			return nil, fmt.Errorf("failed to retrieve image list: %v", err)
+			return nil, errors.Wrap(err, "failed to retrieve image list")
```
Comment (thaJeztah) [inline daemon/images/service.go:253]: Perhaps this should've been in a separate commit, but I guess it's ok

**Review by thaJeztah — APPROVED** [review]
LGTM

**Review by cpuguy83 — APPROVED** [review]

**Pull request merged**